### PR TITLE
v3 high quality CPT  (setting A and B)

### PR DESCRIPTION
**Pull request by YumaTsuta**
This PR adds scripts for the continual pre-trained model of 1.7B-exp2 using high-quality data.

[Issue 22 on llm-jp/experiments](https://github.com/llm-jp/experiments/issues/22)

### CODE REVIEW

**Comment by YumaTsuta** [comment]
# TODO
- [] トークン数の変更

**Comment by YumaTsuta** [comment]
Todo
- [ ] ファイル重複判定
  - [ ] 強制動作オプション

**Comment by YumaTsuta** [comment]
1A, 1B, 2A, 2C, 3A までの設定が完了

**Comment by YumaTsuta** [comment]
@Reviwer
各フォルダーの `dataset_loader.sh`について下記の変更 ( `-d` )を加えた上で、`bash dataset_loader.sh` でデータセットの統計情報が表示されます
``` shell
check_load_dataset -d "$TRAIN_DATA_PATH"
```

**Comment by hkiyomaru** [comment]
実験設定は問題なさそうに見えます

File: pretrain/scripts/v3-high-quality-cpt/1.7B-exp1A/dataset_loader.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+CUR_DIR=${1:-..}
+
+# Common settings
+source ${CUR_DIR}/megatron_data_formatter.sh
+V3_0_INFO_ROOT="/home/shared/corpus/llm-jp-corpus/v3.0.0/training_resharded_tokenize_ver3.0/token_info"
```
[the remark this answers is not in the document]
Comment (YumaTsuta) [reply pretrain/scripts/v3-high-quality-cpt/1.7B-exp1A/dataset_loader.sh:9]: すみません、info fileのコピー依頼中のため
`/home/shared/corpus/llm-jp-corpus` -> `/data/llm-jp-corpus/`
v3.0 token_info: `.sakura_home.csv` ->、`.sakura.csv`
v3.1 token_info: `.sakura_home.csv` ->、`.csv`
でひとまず確認できます。

[the remark this answers is not in the document]
Comment (YumaTsuta) [reply pretrain/scripts/v3-high-quality-cpt/1.7B-exp1A/dataset_loader.sh:9]: @hkiyomaru こちらコピー完了した様なのでレビューの再開お願いします。

**Comment by YumaTsuta** [comment]
@odashi 
Python で書き直しましたレビューお願いします。

＠Reviwer
下記の手順でデータセットの統計情報が表示されます。
```shell
source {path/to/environment}/bin/activate
python3 {path/to/megatron_data_formatter.py} {path/to/data_config.yaml} -d
```
`data_config.yaml`は `pretrain/scripts/v3-high-quality-cpt/1.7B-exp1A/data_config.yaml`などです

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        include_chars (list[str], optional): List of filter patterns. Defaults to an empty list.
+
+    Returns:
+        list[str]: List of formatted dataset strings (token size and data path).
+    """
+    data_paths = []
+    with open(info_file, newline="", encoding="utf-8") as csvfile:
+        reader = csv.reader(csvfile)
+        for row in reader:
+            token_count = int(int(row[2]) * repeat + 0.5)
```
Comment (odashi) [inline pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:59]:
```suggestion
            token_count = round(int(row[2]) * repeat)
```

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        display_details (bool, optional): If True, display detailed token size breakdown. Defaults to False.
+    """
+    total_token_size = 0
+    file_checker = set()
+
+    if display_details:
+        lang_total_tokens: dict[str, int] = {}
+        print(f"{'Lang':<5} {'File Name':<20} {'Token Size':>15}")
+
+    data_entries = train_data_path.strip().split()
```
Comment (odashi) [inline pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:86]:
```suggestion
    data_entries = train_data_path.split()
```

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        token_size = int(data_entries[i])
+        data_path = Path(data_entries[i + 1])
+        lang = data_path.parent.name
+        file_name = data_path.stem
+        combination = f"{lang}/{file_name}"
+
+        if combination in file_checker:
+            common_msg = f"Duplicate entry for {combination}."
+            if not force:
+                raise ValueError(f"{common_msg} Exiting.")
```
Comment (odashi) [inline pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:97]: Exceptionでexitするかどうか決めるのはcatchした側なのでここでは不要
```suggestion
                raise ValueError(common_msg)
```

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            else:
+                print(f"Warning: {common_msg}")
```
Comment (odashi) [inline pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:99]: または明示的に警告を出したいなら `warnings.warn`
```suggestion
                logging.warn(common_msg)
```

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        train_data_path (str): The TRAIN_DATA_PATH string containing dataset paths and token sizes.
+        force (bool, optional): If True, ignore duplicates and continue. Defaults to False.
+        display_details (bool, optional): If True, display detailed token size breakdown. Defaults to False.
+    """
+    total_token_size = 0
+    file_checker = set()
+
+    if display_details:
+        lang_total_tokens: dict[str, int] = {}
+        print(f"{'Lang':<5} {'File Name':<20} {'Token Size':>15}")
```
Comment (odashi) [inline pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:84]: このプログラムの後段の処理で使用しない出力に関しては基本的にloggingを使用してください。他の行についても同様

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        if "basedir" in params:
+            if params["basedir"] in config["common"]:
+                basedir = Path(config["common"][params["basedir"]])
+                assert basedir.exists(), f"Base directory {basedir} does not exist."
+            elif Path(params["basedir"]).exists():
+                basedir = Path(params["basedir"])
+            else:
+                raise ValueError(f"Invalid basedir: {params['basedir']}")
+
+        info_file = basedir / Path(params["file"])
```
Comment (odashi) [inline pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:143]:
```suggestion
        info_file = basedir / params["file"]
```

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        repeat = float(params["repeat"])
+        filters = params.get("filter", None)
+        filters = [filters] if isinstance(filters, str) else filters
+        assert filters is None or isinstance(
+            filters, list
+        ), f"{filters=} must be a string or list of strings."
```
Comment (odashi) [inline pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:149]: ユーザに引数について複数の選択肢を与えるのは避けたほうがよいので、必ずlistを受け取るようにできるならそうしてください。
参考: https://predicatejp.notion.site/Avoid-intelligent-functions-a539b26f384a42708124da551b05aab6

File: pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py (hunk 6 above)
Comment (YumaTsuta) [reply pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py:82]: @odashi loggingを全てstderrにすれば-d オプションいらないので消そうと思いますが、問題ないでしょうか？

Comment (YumaTsuta) [reply pretrain/scripts/v3-high-quality-cpt/megatron_data_formatter.py]: 最終コミットで-dオプション削除したものを作成

**Comment by YumaTsuta** [comment]
@odashi コメント対応しました。
追加でYAMLの仕様を追記しました。

**Comment by YumaTsuta** [comment]
Optimizerを引き継がない前提だったので `--override-opt_param-scheduler`をつけてません。

**Review by odashi — APPROVED** [review]
lgtm
